Drafting utility functions

nmaarnio opened the pull request:
I have now drafted some utility functions and would like your feedback @nialov and @msmiyels . Note that I did not do tests or proper docstrings for all, as many of these might be discarded or modified.

Feel free to shoot any kind of comments and suggestions.

Contents of utilities:
- nodata
  - set_nodata_raster_meta
  - replace_nodata_each_band
  - replace_values_with_nodata
  - replace_nodata_dataframe
  - handle_nodata_as_nan
- file_io
  - get_supported_raster_extensions
  - get_supported_vector_extensions
  - read_spatial
  - read_raster
  - read_vector
  - read_as_raster
  - read_tabular
  - write_raster
  - write_vector
  - write_tabular
- raster
  - raster_profile
  - apply_bandwise
  - separate_bands (not implemented)
  - merge_rasters (not implemented)

EDIT: Thinking more about processing raster data, I think the _apply_bandwise_ is not sufficient. Now it reads all bands, but in many cases as @msmiyels pointed out we want to process only some bands.

nialov commented:
I should be able to get a review with suggestions in this week. I will comment here when I get started!

nialov commented:
Will take a look now.

nialov commented on eis_toolkit/checks/dataframe.py:
This kind of check should, instead of a boolean, return the columns that are not numeric. It is very possible that `pandas` misinterprets numeric data as strings. It is quite easy in those cases to call `df[<col>].astype(float)` to convert to numeric data. This could be tried for the columns that are not numeric before raising an exception for the user. I would also consider expanding the check functions to raise the error itself instead of outside the check function. If in some function we do not want to instantly raise an exception, it can be caught as the check functions probably have it well annotated which exception they are going to raise (e.g. `InvalidParameterValueException`).

nmaarnio replied:
How often do we need to stop execution because there is a non-numeric column, instead of just filtering correct columns (and raise exception if 0 columns are numeric)? I am bit unsure how often this function would be needed, but we can implement this if you feel it is needed.

nialov commented on eis_toolkit/checks/raster.py:
The if-else statements are unnecessary. Can just do:

~~~python
return abs(raster.transform[0]) == abs(raster.transform[4])
~~~

and

~~~ python
return all(band in range(1, raster.count + 1) for band in bands)
~~~

nmaarnio replied:
There is actually nothing new in this file, these checks were implemented by @Mtk112 when we needed a functionality to check if raster grids are equal. The subtasks were separated into their own checks. I think I have simply forgotten the `NonSquarePixelSizeException`, it was previously used in `snap_raster` function but removed later as unnecessary. But we could add these two checks you suggested.

nialov commented on eis_toolkit/utilities/raster.py:
Wrapping functions like so causes the function arguments and outputs to not be available for e.g. code completion but I suppose this is mostly for internal use so not too worrysome. You should also do `yield function(band_data, *args, **kwargs)` so the results are returned through a generator. As @msmiyels said these are probably not worth focusing on currently.

nialov commented on eis_toolkit/utilities/nodata.py:
```diff
@@ -0,0 +1,119 @@
+    Returns:
+        out_data: Raster data with updated nodata values.
+    """
+    out_data = data.copy()
+    out_data[np.isinf(out_data)] = new_nodata  # Is this line needed?
+    mask = np.isin(data, values_to_replace)
+    out_data[mask] = new_nodata
+    return out_data
```
```suggestion
    return np.where(np.isin(data, values_to_replace) | np.isinf(data), new_nodata, data)
```

nmaarnio commented on eis_toolkit/checks/argument_types.py:
```diff
@@ -0,0 +1,29 @@
+from eis_toolkit.exceptions import InvalidArgumentTypeException
+
+
+def check_argument_types(func: Callable) -> Callable:
```
Ok, I will give `beartype` a try next.

nmaarnio commented on eis_toolkit/checks/dataframe.py:
```diff
@@ -0,0 +1,18 @@
+import pandas as pd
+
+
+def check_columns_valid(df: pd.DataFrame, columns: Iterable[str]):
```
Sounds good to me.

nmaarnio commented on eis_toolkit/utilities/file_io.py:
```diff
@@ -0,0 +1,112 @@
+from eis_toolkit.exceptions import InvalidParameterValueException, UnsupportedFileExtensionException
+
+
+def get_supported_raster_extensions() -> List[str]:
+    """Get raster file extensions supported by rasterio."""
+    extensions = list(rasterio.drivers.raster_driver_extensions().keys())
+    return extensions
+
+
+def get_supported_vector_extensions() -> List[str]:
+    """Get vector file extensions supported."""
+    extensions = ["shp", "geojson", "gpkg", "csv", "kml", "gml"]
+    return extensions
+
+
+def read_spatial(file_path: Path) -> rasterio.io.DatasetReader | gpd.GeoDataFrame:
+    """Read and open a vector or raster file."""
+    file_extension = file_path.split(".")[-1]
+    if file_extension in get_supported_vector_extensions:
+        data = rasterio.open(file_path)
+    elif file_extension in get_supported_vector_extensions():
+        data = gpd.read_file(file_path)
+    else:
+        raise UnsupportedFileExtensionException(
+            f"File extension {file_extension} is not supported. Provide a raster or vector file."
+        )
+    return data
```
Good points @nialov . I'll think about this a little bit. I can imagine it was thought in the beginning that classes won't be needed, but we are not bound by the initial decisions as now we know more about the toolkit and our options.

nmaarnio replied:
If I'm not mistaken, `list(rasterio.drivers.raster_driver_extensions().keys())` should cover all formats supported by rasterio, but a similar listing does not exist for vector data. Did you have something like this as an alternative in your mind Nikolas?

```
def read_spatial(file_path: Path):
    try:
        data = read_raster(file_path)
    except rasterio.errors.RasterioIOError:
        try:
            data = read_vector(file_path)
        except (ValueError, OSError):
            try:
                data = read_csv(file_path)
            pd.errors.ParserError:
                raise FileReadError("Unsupported file format or corrupted file")

    return data
```

Here I'm not sure if all the possible relevant exceptions are caught.

nialov replied:
Yes, something like that. It will probably require some trial and error to catch all exceptions. However, just as a side note, I would not consider this functionality of accepting tabular and geometry data in place of rasters a priority in terms of time. We should focus on making everything work with rasters first and then start adding the additional formats.

nmaarnio commented on eis_toolkit/utilities/nodata.py:
```diff
@@ -0,0 +1,119 @@
+    Returns:
+        raster_meta: Raster metadata with updated nodata value.
+    """
+    raster_meta.update({"nodata": nodata_value})
```
Good reminder, forgot about copying. However, wouldn't raster_meta.copy() suffice? This shallow copying is done in the existing raster processing functions.

nialov replied:
Probably, I just tend to use `deepcopy` to avoid the rare copying bugs.

nialov commented on eis_toolkit/utilities/nodata.py:
```diff
@@ -0,0 +1,119 @@
+        out_data = replace_values_with_nodata(result, np.nan, nodata_values)
+        return out_data
+
+    return wrapper
```
I think all downstream functions can probably expect that the nodata is already handled (i.e. replaced with `np.nan`). I cannot readily say when we would have functions that require the explicit value of nodata from the raster instead of `np.nan`. So this decorator would mostly be in use close to the command-line so the implementation details are probably up to the person who does the command-line.

nmaarnio commented:
I have made some modifications to the functions and removed some fully. I think it would be good to get some batch merged soon, and it may take more time to really figure out what we want from some functions. What remains now in the PR is:
- nodata functions (with small updates) with basic tests
- updated reading functions (although I don't think these need to be merged right now, so we can scratch these too)
- some checks (added some according to the dicsussion)
- added beartype and tested it (beartype causes now these GH tests to fail, though)

Even though it's important to have these utility functions, many of them are needed between the CLI and the data functions which became more clear through our discussion. These functions that live close to CLI (reading, writing, basic nodata handling, doing input data selections (based on selections made in QGIS UI or via CLI)) shouldn't be needed in implementing the data functions. So, maybe it is easier to postpone implementing them a little bit.

Regarding changing checks to raise exceptions, I would not do that in this PR but in a refactoring PR. However, thinking about this matter again, I see value in having the checks always return a boolean: This way the exceptions are transparently raised in the toolkit functions, and I think there are still many cases where a check would be used to compress and reuse if statements.

Do you have any quick comments about what is now included in the PR? While these utility functions are important, I don't want to take too much of your (or my) time to bounce these back and forth either.

nialov commented:
> The only point not clear for me in the nodata replacement is
> 
> ```python
> return np.where(np.isin(data, values_to_replace) | np.isinf(data), new_nodata, data)
> ```
> 
> `np.inf` can also be part of the `values_to_replace` when needed. We also need to keep in mind that this only works for arrays with **floating** dtypes.

This was drafted by meand you are correct, np.inf can just be added to `values_to_replace` so the `| np.isinf(...)` is not needed!

nialov commented:
You can just try adding `beartype` to `environment.yml` and see if the conda
workflows work.

~~~patch
diff --git a/environment.yml b/environment.yml
index c5eacc3..66cae92 100755
--- a/environment.yml
+++ b/environment.yml
@@ -15,5 +15,6 @@ dependencies:
   - statsmodels >=0.13.5,<1.0.0
   - keras >=2.10.0,<3.0.0
   - tensorflow >=2.10.0,<3.0.0
+  - beartype >=0.13.1,<2.0.0
   # Dependencies for testing
   - pytest >=7.2.1
~~~

nialov commented:
You can also update the `requirements.txt` with

~~~bash
poetry export --without-hashes > requirements.txt
~~~

nmaarnio commented:
Thanks for the comments! I...

- created two specific np.nan conversion functions
- fixed the nodata decorator
- removed the np.isinf
- added miscellaneous utilities that now has generic replacement functions (as I created the np.nan functions, I thought we might as well actually make the replacement functions more generic)
- updated requirements.txt and environment.yml
- reverted to Union in type hints (I could not get "|" working in GH actions. Maybe there is a workaround, let me know if you find it.)
- some type hints from Beartype (Beartype informed that some type hints will be deprecated in the future and suggested this to silence its warnings)

What do you think now? I can remove the file I/O functions if we think they are not ready – we won't need them anyway for the data functions, only when we/I develop the CLI.

nialov commented on eis_toolkit/utilities/miscellaneous.py:
```diff
@@ -0,0 +1,71 @@
+    Returns:
+        Raster data with replaced values.
+    """
+    if copy:
+        out_data = data.copy()
+    else:
+        out_data = data
```
`np.where` always returns a new array so the input array is never modified and these lines do therefore nothing. I do not see any reason why someone would want to modify the original input data directly i.e. have a function with side-effects.

nialov reviewed (approved):
I do not see any need to polish these functions further at this point. Most are or at least should be only relevant when implementing the command-line. They can be modified later.
